zilliqa/src/p2p_node.rs

```diff
@@ -432,6 +435,28 @@ impl P2pNode {
                         break;
                     }
                 }
+                () = &mut sleep => {
+                    let net_info = self.swarm.network_info();
+                    trace!("p2p_node tick {0} / {1} ", net_info.num_peers(), net_info.connection_counters().num_connections() );
+                    if net_info.num_peers() == 0 && net_info.connection_counters().num_connections() == 0 {
+                        // We have no peers and no connections. Try bootstrapping..
+                        if let Some((peer, address)) = &self.config.bootstrap_address {
+                            if self.swarm.local_peer_id() == peer {
+                                debug!("p2p_node: can't bootstrap against myself");
+                            } else {
+                                debug!("p2p_node: no peers and no connections - bootstrapping!");
+                                self.swarm
+                                    .behaviour_mut()
+                                    .kademlia
+                                    .add_address(peer, address.clone());
+                                self.swarm.behaviour_mut().kademlia.bootstrap()?;
```

Comment:
z2 doesn't work without this change, or I wouldn't've had to spent two days finding out we needed it :-)

The logic here is that, at least at startup, a node starts before the bootstrap; it fails to add the peer (because it's not up yet), and drops the bootstrap. Now it won't add the peer again, even if it comes up, and z2 then deadlocks essentially forever.

I'll see if `automatic_bootstrap_throttle` will work, but if it's currently set to less than a few hours, I suspect it is already triggering. We could perhaps omit the `.kademlia_bootstrap()`, but the `.add_address()` definitely isn' optional.

Reply:
@JamesHinshelwood  - OK. I've checked. You can't get away without re-adding the address, but (as expected) kademlia will eventually bootstrap itself automatically.